components/core/src/crypto/keys/cache.rs

```diff
@@ -0,0 +1,515 @@
+use crate::{crypto::{hash,
+                     keys::{encryption::{generate_origin_encryption_key_pair,
+                                         generate_service_encryption_key_pair,
+                                         generate_user_encryption_key_pair},
+                            generate_signing_key_pair,
+                            KeyFile,
+                            NamedRevision,
+                            OriginPublicEncryptionKey,
+                            OriginSecretEncryptionKey,
+                            PublicOriginSigningKey,
+                            RingKey,
+                            SecretOriginSigningKey,
+                            ServicePublicEncryptionKey,
+                            ServiceSecretEncryptionKey,
+                            UserPublicEncryptionKey,
+                            UserSecretEncryptionKey}},
+            error::{Error,
+                    Result},
+            fs::AtomicWriter};
+use std::{convert::TryFrom,
+          io::Write,
+          path::{Path,
+                 PathBuf}};
+
+/// Represents the location of all Habitat keys (user, service,
+/// origin, signing, and ring) locally on disk, as well as the APIs
+/// for retrieving and storing keys.
+#[derive(Clone, Debug, PartialEq)]
+pub struct KeyCache(PathBuf);
+
+impl AsRef<Path> for KeyCache {
+    /// Expose the path to this key cache.
+    fn as_ref(&self) -> &Path { self.0.as_ref() }
+}
+
+impl KeyCache {
+    pub fn new<P>(path: P) -> Self
+        where P: Into<PathBuf>
+    {
+        KeyCache(path.into())
+    }
+
+    /// Ensure that the directory backing the cache exists on disk.
+    pub fn setup(&self) -> Result<()> {
+        if !self.0.is_dir() {
+            std::fs::create_dir_all(&self.0)?;
+        }
+        Ok(())
+    }
+
+    ////////////////////////////////////////////////////////////////////////
+
+    /// Generate a new ring key and save it to disk.
+    pub fn new_ring_key(&self, name: &str) -> Result<RingKey> {
+        let key = RingKey::new(name);
+        self.write_key(&key)?;
+        Ok(key)
+    }
+
+    /// Generate a new origin signing key pair and save both keys to disk.
+    pub fn new_signing_pair(&self,
+                            origin: &str)
+                            -> Result<(PublicOriginSigningKey, SecretOriginSigningKey)> {
+        let (public, secret) = generate_signing_key_pair(origin);
+        self.write_pair(&public, &secret)?;
+        Ok((public, secret))
+    }
+
+    /// Generate a new origin encryption key pair and save both keys to disk.
+    pub fn new_origin_encryption_pair(
+        &self,
+        origin: &str)
```

Comment:
Good call; will do.

Reply:
:thinking: This isn't hard, but it is a bit tedious and will touch a number of files. This might be a good candidate for a follow-up PR, since this PR is already a bit large.